### ISSUE TITLE
[bug]: screen readers have a hard time with the UI

### ISSUE THREAD

**Issue by zersiax** [makeplane/plane#1154]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

I had a quick look around the Plane interface and while a lot o things work, marginally, well, there's also a lot of things that could do with some serious accessibility improvements.
Many form fields lack an accessible name/label, heading structure is all over the place, several buttons are unlabeled for assistive tech etc.
Is there a plan to tackle the accessibility of the interface so people don't get fired when a company decides to adopt this tool only to find out that some of their employees can now no longer keep track of the projects they're working on?

### Steps to reproduce

Test using the interface with only the keyboard , while a screen reader like [NVDA](https://nvaccess.org).

### Browser

Google Chrome

### Version

Cloud

**Comment by aaryan610** [comment]
Hi @zersiax, thank you for highlighting the accessibility concerns. We want to emphasize that we are actively working on enhancing keyboard accessibility as well. We recognize the importance of keyboard navigation for users who rely on it and are committed to providing a seamless experience.

Your feedback is greatly appreciated as we strive to make Plane more inclusive and user-friendly. Thank you for your support and patience as we continue to make improvements.

**Comment by zersiax** [comment]
Please note that keyboard accessibility and screen reader accessibility are not necessarily equivalent. Keyboard accessibility can complement screen reader accessibility, but will not cover both cases.

**Comment by vihar** [comment]
@zersiax, it would be greatly appreciated if you could suggest any frameworks or libraries that we could utilize for implementing this, or provide ideas if we need to develop it from scratch.

@Shivangi-jain33, @bhaveshraja, let's include this issue in our roadmap.

**Comment by zersiax** [comment]
This is not really a problem you solve by using a library, for the most part. While design systems and component libraries exis that are pretty accessible out of the box, what it reallly comes down to a lot of the time is the HTML that ultimately gets rendered by whatever framework you're using. If a button a user clicks visually looks like a button, but is really just a div with a whole bunch of JS attached to it to make it act like a button, then you're either going to have to turn that div into an actual button element, or manually add the required aria properties (aria-label, role, tabindex etc) to make it behave like a button. If a button only has an icon on it and no text, a screen reader isn't going to know how to read out its function for its users, so you'd need an aria-label which sets that button's accessible name.
Tl,dr: Use semantic HTML and 80% of your problems disappear.

**Comment by zersiax** [comment]
Look for example at the input.tsx in ui/form-fields.
In HTML, there's two ways to attach a label to a form field. THe label element with a for-attribute that points to the ID of the label, or an aria-label or aria-labelledBy on the input element itself. I don't see a prop for aria-label on this component, and I don't see a label UI element in that folder, maybe there's one in components but I don't really have the time to dig deep right now. 
Therefore, A supersimple fix could be to add aria-label as a prop on input and textarea, make sure that components that call this widget supply a value to it and now your form fields are labeled.
A button element generally gets its accessible name from it's inner text, but if there isn't any, you'd have to apply an aria-label there as well. Same for checkboxes, links with just an icon in them, etc.
Things like dropdowns, modals and autocompletes are harder, but this'd be a show of commitment already.

**Comment by rzr** [comment]
To improve accessibility, may I suggest to make a minimalist view that just rely on default platform behavior (eg: HTML2 widgets, without JS overrides)

ps: I don't use a screen reader and I am not visually impaired but i am struggling to use the UI, I am considering to make a wrapper on REST to make my life easier.

**Comment by shindere** [comment]
Many thanks to @zersiax for having raised the concern and to @rzr who has pointed this issue to me

## Personal context

I am a completely blind software engineer. I work under Linux, mostly in text-mode (by which I mean Linux's virtual console, not a graphical terminal) with the exception of Firefox that I use for the web because most of nowadays websites simply cannot be used any longer with a text-mode or Javascript-unaware browser.

The screen reader that is used under Linux is called Orca. It renders the screen content through both text to speech and in braille, provided that the user has a braille device connected to their computer.

## Accessibility background

The W3C has developed the [Web Content Accessibility Guidelines](https://www.w3.org/TR/WCAG21/) which give a set of recommendations to make the web accessible to everybody, regardless of their situation or accessibility needs.

## Keyboard accessibility

As has been rightly explained, although being able to interact with the website without using a mouse certainly is part of what needs to be achieved, it can't be enough. Indeed, if every button can be accessed with tab, for example, but all you know about that button is that it's a "button" (because no descriptive label has been attached to it), then the website will not really be accessible for a visually impaired person.

So if a developer really wanted to ensure that something is accessible, then they should not only refrain from using their mouse. They should also install a screen reader and, once on the website, they should close their eyes and try to perform their task just by listening to the speech synthesis and using the keyboard. And to make sure such an experiment is conclusive, also make sure to log out before you start the experiment so that you can also check how well the log-in process itself works.

That being said, a word of caution is due here: a screen reader is like any other computer program, it needs a bit of time and efforts to learn how to use it. But, I think, if one's aim is to become aware of others' difficulties and thus to become a more mindful and inclusive developer, then it it's worth it. And I will be happy to support whoever is willing to engage by referring to appropriate resources depending on the operating system you are using.

## It's not about the library

As has also been rightly emphasised, it's really not a matter of library. Well, it can be, but you can produce inaccessible pages even with a library that supports accessibility well if you do not take the time to properly label your buttons or decide which of your pictures are decorative (and should thus be hidden to assistive technologies by giving them an *empty* alternative text) or which carry information (and should thus have a proper, *non-empty* alternative text associated to them).

To some extent, the question of what library to use to build accessible websites could be compared to the question: what library to use to write clean or secure code. In these two examples, the answer may be something like well, it's not so much a question of the library, it's more a question of culture and understanding what needs to be done or avoided and why. The same equally applies to accessibility.

## And now, what?

Ideally, the website should be audited for accessibility. The output of such an audit is a detailed report of what is not accessible yet and what should be done to make it accessible. I should clarify here that the outcome of an audit is not only a list of problems you are left alone with. The output of an audit is a list of problems, yes, with their consequences on users explained, and also a list of actionable things you can do to remedy these problems. Possibly the report will list the issues by impact (which issues impact most of the users) and / or by complexity of resolution.

So there may be some quick wins: issues that both impact many users and are easy to fix. Yes, that happens, too, and is very rewarding for developers.

Once accessibility has been achieved, maintaining it while the website changes requires the developers to receive proper training and support so that writing accessible applications becomes second nature to them, just as writing clean, factorised, secure and easy-to-maintain code.

As a side-note, accessible websites and well-structured code often go hand in hand because making code accessible means introducing more semantics in the pages, which then makes them easier to evolve. For instance, instead of relying on the visual convention that all your titles will be in blue and underlined and then having to manually underline and put in blue each title, you define a CSS class for them stating that they are titles and you then write it once and for all that the members of the title class need to be in blue and underlined. The day you want to change it, you change it in one place and everything remains consistent. As simple as this example may feel, it has the virtue of being easily understandable by everybody. There are actually> many other, way more involved examples illustrating how structuring the code well and making the corresponding interface accessible are, so to speak, isomorphic to each other.

## Commercial value of all this

Beyond the ethical imperative, there are also practical reasons to prioritise accessibility.

More and more governments and public agencies do prioritise accessible solutions over non-accessible ones. So a proactive and continuous commitment towards accessibility will likely become a more and more compelling differentiator.

In Europe, for instance, there is the [European Accessibility Act](https://commission.europa.eu/strategy-and-policy/policies/justice-and-fundamental-rights/disability/european-accessibility-act-eaa_en) which is in place since June 2025. Given this legal framework, it is reasonable to expect that more and more actors will not even take inaccessible (or partially accessible) solutions into account, no matter what merits they may have otherwise. So, the sooner one is prepared, the better it will be.

**Comment by zersiax** [comment]
I honestly haven't had a chance to look at the interface since i filed this ticket 3 years ago, has anything been done since then to improve the situation?
Happy to take another look if people are still willing to fix this.

**Comment by rzr** [comment]
Maybe we can start by identifying "features" that could be made optional ?

- [ ] Avoid Platform overrides (shortcuts etc)
  - https://github.com/makeplane/plane/issues/8663
- [ ] Single content page option (no side pane in list views, prefer to open link in separate tabs)
